From 180b91dcc449541c4f15d6095a43b258caceb3cb Mon Sep 17 00:00:00 2001
From: Jan Sondhauss <jan.sondhauss@wago.com>
Date: Mon, 14 Oct 2019 11:05:37 +0200
Subject: [PATCH] driver: net: dsa: ksz8863: fix port seperation

In the seperated mode the learning and tx must be disabled
for the external ports.

Signed-off-by: Jan Sondhauss <jan.sondhauss@wago.com>
---
 drivers/net/dsa/ksz886x/ksz8863.c | 22 +++++++---------------
 1 file changed, 7 insertions(+), 15 deletions(-)

diff --git a/drivers/net/dsa/ksz886x/ksz8863.c b/drivers/net/dsa/ksz886x/ksz8863.c
index 1963903..9a0fedea8 100644
--- a/drivers/net/dsa/ksz886x/ksz8863.c
+++ b/drivers/net/dsa/ksz886x/ksz8863.c
@@ -435,7 +435,6 @@ static int ksz8863_set_bridged_mode(struct ksz8863_chip *chip)
 	int err, ret;
 	int (*set_mode)(struct ksz8863_chip *);
 	int port_index;
-	u8 states[2];
 
 	set_mode = NULL;
 
@@ -453,19 +452,6 @@ static int ksz8863_set_bridged_mode(struct ksz8863_chip *chip)
 
 	ksz8863_reg_lock(chip);
 
-	/* backup port register ctrl 2 */
-	for (port_index = 0; port_index < KSZ8863_NUM_PORTS; port_index++) {
-		int pb = KSZ8863_REG_PORTS_BASE * port_index;
-
-		if (dsa_is_cpu_port(chip->ds, index_to_port(port_index)))
-			continue;
-
-		err = ksz8863_mii_read(chip, pb + KSZ8863_REG_PORT1_CTRL2,
-				       &states[port_index]);
-		if (err)
-			goto out_unlock;
-	}
-
 	/* disable learning and disable receiving while changing
 	 * handling of unknown mac address and broadcast frames
 	 */
@@ -489,12 +475,18 @@ static int ksz8863_set_bridged_mode(struct ksz8863_chip *chip)
 	/* restore port register ctrl 2 */
 	for (port_index = 0; port_index < KSZ8863_NUM_PORTS; port_index++) {
 		int pb = KSZ8863_REG_PORTS_BASE * port_index;
+		u8 state = KSZ8863_REG_PORT_CTRL2_RX_ENABLE;
 
 		if (dsa_is_cpu_port(chip->ds, index_to_port(port_index)))
 			continue;
 
+		if (!chip->switched)
+			state |= KSZ8863_REG_PORT_CTRL2_LEARN_DISABLE;
+		else
+			state |= KSZ8863_REG_PORT_CTRL2_TX_ENABLE;
+
 		err = ksz8863_mii_write(chip, pb + KSZ8863_REG_PORT1_CTRL2,
-					states[port_index]);
+					state);
 		if (err)
 			goto out_unlock;
 	}
-- 
2.7.4

